add WEffectMetaKnob, draws arc from default meta position

ronso0 opened the pull request:
When an effect is loaded, the knob gets the default meta parameter from the effect manifest and draws the knob arc from there to current angle.

Closes #12634

Don't let this distract you from the 2.4 release, this can very well wait for 2.4.1 or later.

I added this widget only where arcs are or may be used, i.e. all skins except Shade.

ronso0 committed:
new WEffectMetaKnob: knob that draws arc from dynamic default position

ronso0 committed:
add WEffectMetaKnob to skins (arcs are drawn only in LateNight PaleMoon)

daschuer commented:
This one looks now: 
![grafik](https://github.com/mixxxdj/mixxx/assets/1777442/2050731a-429d-4f06-8cdf-4f618280bd46)
The Meta knob has the correct ring, but the parameter knob is still wrong.

ronso0 committed:
WEffectParameterKnobComposed: draw arc from neutral point on scale, default 0

ronso0 committed:
effect parameters: fix/remove various neutral-point-on-scale values

ronso0 commented:
No prob, ~~fix~~ second new feature is here.

This revealed some misplaced 'neutral point on scale' values, e.g. high freq. of Graphic EQ.

ronso0 commented:
Draft so I can clean up and force-push  without disrupting a review.

Swiftb0y commented on src/widget/wknobcomposed.cpp:
```diff
@@ -116,6 +114,15 @@ void WKnobComposed::onConnectedControlChanged(double dParameter, double dValue)
+    if (parameter < 0 || parameter > 1) {
+        m_defaultAngleIsValid = false;
+    } else {
+        m_defaultAngle = m_dMinAngle + (m_dMaxAngle - m_dMinAngle) * parameter;
```
```suggestion
        m_defaultAngle = std::lerp(m_dMinAngle, m_dMaxAngle, parameter);
```

ronso0 replied:
nice one, thanks!

Swiftb0y commented on src/effects/effectknobparameterslot.cpp:
```diff
@@ -239,3 +239,10 @@ void EffectKnobParameterSlot::syncSofttakeover() {
+    if (m_pManifestParameter) {
+        return m_pManifestParameter->neutralPointOnScale();
+    }
+    return 0.0;
```
this looks like a std::optional candidate. wdyt?

daschuer replied:
In this case 0.0 is a good default and valid value for the unlike case that the is no parameter connected.

Swiftb0y replied:
but 0.0 could also be a valid neutral point (for example in a high-pass filter), it's impossible for the caller to tell. I think the caller needs to explicitly handle the circumstance (for example by showing "error" as a value, not just 0).

daschuer replied:
An error value for the neutral position, when we have also no value and no scale? 
This case should never happen, since the parameter knobs are hidden in that case. So coding around that is not reasonable. 

We have that state for the meta knob when no effect is loaded. Currently a 0 .. 1 knob with default is assumed and no error state. Since the value of the knob remains where is was when loading "---".  What kind of error state you have in mind?

Swiftb0y replied:
> An error value for the neutral position, when we have also no value and no scale?

> What kind of error state you have in mind?

The error state is if `m_pManifestParameter == nullptr`. 

> We have that state for the meta knob when no effect is loaded. Currently a 0 .. 1 knob with default 0 is assumed and no special error state. The value of the knob remains where is was when loading "---".

That doesn't seem like particularly good design IMO because that assumes some other part of the code does the appropriate checking. This assumption could get violated easily in a future refactoring. If you just look at this public member function (thus part of the class's API), there is no way to verify whether the value is returned represents an error or real value. And even if there is, the caller would have to know about that implementation detail. The result is an API that is easy to misuse.

Swiftb0y replied:
> Are you satisfied with 1. ?

Yes. 

>  In this particular case it does not makes sense to consider a case that never happens.
We can probably ASSERT that, I am just afraid that will bite us later.

playing devils advocate here, this sounds nonsensical. If it never happens, then assert and encode that in the type-system (plain double, no special values indicating an error state). If it happens, lets encode it in the type system (`std::optional<double>`).

Swiftb0y commented on src/widget/wknobcomposed.h:
```diff
@@ -28,6 +28,10 @@ class WKnobComposed : public WWidget {
     void paintEvent(QPaintEvent* /*unused*/) override;
 
+    bool m_defaultAngleIsValid;
+    double m_defaultAngle;
```
if you wrap this in an optional, its impossible to accidentally use it without checking.

ronso0 replied:
so `std::optional<double> m_defaultAngle;` would not require ..isValid anymore, we'd just check
`if (m_defaultAngle) { .. } else ..`

ronso0 replied:
No, I'm not getting it.
(your comment in _this_ place)

I'll try using std::optional<double> for neutralPointOnScale()

Swiftb0y replied:
yes, though its often nicer to use `value_or` and the monadic operations added in C++23. Id also prefer `has_value` over `operator bool` in if conditions. [`std::optional<T>::operator->, std::optional<T>::operator*`](https://en.cppreference.com/w/cpp/utility/optional/operator*) should be avoided!

ronso0 replied:
ah okay, I was confused why I could use m_angle right away, std::optional<T> has `value()`

Swiftb0y replied:
yeah. the difference is that one is checked the other is not. The problem with that API is that the check could now be very far away from the use and could get lost in a future refactoring. `value()` is safer (in the sense that it `std::terminates` when there is no value, instead of UB) but obviously checks each and every time. optional monadic operations `C++23` solve that.

Swiftb0y reviewed (commented):
just brainstorming here. not trying to drag this out.

daschuer reviewed (approved):
Works perfect. Thank you.

ronso0 committed:
Effects/GUI: use std::opt<double> for neutralPointOnScale()

Swiftb0y commented on src/widget/weffectmetaknob.h:
```diff
@@ -0,0 +1,23 @@
+    void effectChanged();
+
+  private:
+    EffectsManager* m_pEffectsManager;
+    EffectSlotPointer m_pEffectSlot;
```
encode constraints and ownership in the type system (`gls::not_null<EffectsManager*>` would blow the scope too much right now). implicit info should at least receive a comment.
```suggestion
    gls::not_null<std::weak_ptr<EffectsManager>> m_pEffectsManager;
    // m_pEffectSlot can be null, in which case the effect does not have an effect assigned.
    EffectSlotPointer m_pEffectSlot;
```

ronso0 replied:
Ahmm, and change it everywhere in src/effects ??

There is `VERIFY_OR_DEBUG_ASSERT(m_pEffectSlot)` in
`WEffectMetaKnob::setup` and
`WEffectParameterKnobComposed::setup`
and in case of null the respective slots where `m_pEffectSlot`/`m_pParameterSlot` are accessed are not connected.

Swiftb0y replied:
> Ahmm, and change it everywhere in src/effects

In the long term, yes. Not in this PR.

> There is VERIFY_OR_DEBUG_ASSERT(m_pEffectSlot)

Yes, but only on setup. The raw pointer could become dangling, thus a `std::weak_ptr` to ensure that can't happen (and to prevent ref cycles, which would likely form if it was replaced by `std::shared_ptr`. 

Don't get me wrong, the current solution works, no doubts, it's just not very robust IMO.

Swiftb0y commented on src/widget/weffectmetaknob.cpp:
```diff
@@ -0,0 +1,46 @@
+
+    if (pManifest) {
+        setDefaultAngleFromParameterOrReset(pManifest->metaknobDefault());
+        setBaseTooltip(QString("%1\n%2").arg(
```
```suggestion
        setBaseTooltip(QStringLiteral("%1\n%2").arg(
```

Swiftb0y commented on src/widget/weffectmetaknob.cpp:
```diff
@@ -0,0 +1,46 @@
+                pManifest->name(),
+                pManifest->description()));
+    } else {
+        // This makes drawArc() fall back to drawing from center (if Reverse is not set)
+        setDefaultAngleFromParameterOrReset(-1);
```
seems better now that we take an optional.
```suggestion
        setDefaultAngleFromParameterOrReset(std::nullopt);
```

ronso0 replied:
:+1: thanks, I overlooked that

Swiftb0y commented on src/widget/wknobcomposed.cpp:
```diff
@@ -116,6 +114,14 @@ void WKnobComposed::onConnectedControlChanged(double dParameter, double dValue)
 }
 
+void WKnobComposed::setDefaultAngleFromParameterOrReset(std::optional<double> parameter) {
+    if (!parameter.has_value() || parameter.value() < 0 || parameter.value() > 1) {
+        m_defaultAngle = std::nullopt;
+    } else {
+        m_defaultAngle = std::lerp(m_dMinAngle, m_dMaxAngle, parameter.value());
+    }
```
not a suggestion, just a preview of what would be possible in C++23. (imo easier to read and also faster because it only does a single has_value check. 
```suggestion
    m_defaultAngle = parameter.transform([this](double value) {
        return (parameter < 0 || parameter > 1) ? 
            std::nullopt : 
            std::lerp(m_dMinAngle, m_dMaxAngle, parameter);
    });
```

Swiftb0y replied:
> In general I would prefer to give m_defaultAngle a valid value here. It is better to have a variable that can be used unconditional in all cases.

That defeats the purpose of the optional. This snippet of code here can't accurately determine what value is appropriate.

Swiftb0y reviewed (commented):
looks like ur currently missing a couple `#include <optional>`s.

ronso0 committed:
Effect widgets: use QStringLiteral for tooltip construction

daschuer commented on src/widget/weffectparameterknobcomposed.cpp:
```diff
@@ -25,13 +25,16 @@ void WEffectParameterKnobComposed::setup(const QDomNode& node, const SkinContext
+        setBaseTooltip(QStringLiteral("%1\n%2").arg(
                 m_pEffectParameterSlot->name(),
                 m_pEffectParameterSlot->description()));
+        neutralPoint = m_pEffectParameterSlot->neutralPointOnScale();
```
If we have here a std:nullopt we have a skin error. Likely the a button slot is connected to a knob. 
Should we issue a warning here?

daschuer replied:
Ok, I have not tested that case. That is sufficient than.

daschuer commented on src/widget/weffectmetaknob.cpp:
```diff
@@ -0,0 +1,45 @@
+                pManifest->name(),
+                pManifest->description()));
+    } else {
+        setDefaultAngleFromParameterOrReset(std::nullopt);
```
This means that the "ArcUnipolar" and "ArcReversed" take control which are not set from the skin. The old position remains. This is not consistent. I would prefer to also keep the old Arc. 
We may introduce a "not in use state" in a follow up PR as suggested form @Swiftb0y

ronso0 replied:
If a skin designer unsets `<ArcUnipolar>` and/or sets `ArcReversed` for a Meta knob or effect parameter knob, well, then there is _some_ intention behind it :thinking: :man_shrugging:

daschuer replied:
Since this PR, the default position is taking over. I consider adopting the old values for the "---" selections as a minor inconsistency.

Swiftb0y reviewed (approved):
LGTM. feel free to force-push the squashed fixups. Thanks for your patience.

Swiftb0y commented on src/widget/wknobcomposed.cpp:
```diff
@@ -116,6 +114,14 @@ void WKnobComposed::onConnectedControlChanged(double dParameter, double dValue)
 
+void WKnobComposed::setDefaultAngleFromParameterOrReset(std::optional<double> parameter) {
+    if (!parameter.has_value() || parameter.value() < 0 || parameter.value() > 1) {
+        m_defaultAngle = std::nullopt;
```
whats the advantage here?

ronso0 replied:
Yes, we could have just one painter call, though at the cost of more temp vars.

And there's some more potential for optimization, sure.
For exampe the rect doesn't need to be calculated each time drawArc() is called, it'd be sufficient to do this in a resize event (which is called on initial show() and when being resized of course, if someone prefers to use variable sizes)

ronso0 commented on src/widget/weffectparameterknobcomposed.cpp:
```diff
@@ -25,13 +25,16 @@ void WEffectParameterKnobComposed::setup(const QDomNode& node, const SkinContext
         // indicates no parameter is loaded, so this tooltip should never be shown.
         setBaseTooltip("");
     }
+    setDefaultAngleFromParameterOrReset(neutralPoint);
```
The knob doesn't know if it's visible (sure, in official skins it's hidden if there's no parameter loaded)

daschuer replied:
In the theoretical case it is not hidden, the same ideas like the meta knob applies. 
It does not make sense to change the scale to center. Just keeping it unchanged seems to be consistent. 

Or remove the arc entirely or such in a follow up PR.

ronso0 committed:
WKnobComposed: calculate arc rectangle only when the knob is resized

daschuer commented:
By the way: Is this a last minute 2.4 PR? 
I don't mind. At least my latest comments are not intended blocking a merge.

ronso0 commented:
Yes, that can slip into 2.4.1 last minute!
(need to double-check all skins for regressions! though I need to quit now, can take care of it on sunday earliest)

Please check the rect commit and the fixup with your suggestion.

daschuer commented:
> Please check the rect commit and the fixup with your suggestion.

Looks good. Thank you.

ronso0 commented:
> Error: ../src/widget/wknobcomposed.cpp:118:45: error: 'value' is unavailable: introduced in macOS 10.13
    if (!parameter.has_value() || parameter.value() < 0 || parameter.value() > 1) {
                                            ^
/Applications/Xcode_13.2.1.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX12.1.sdk/usr/include/c++/v1/optional:954:27: note: 'value' has been explicitly marked unavailable here
    constexpr value_type& value() &
                          ^

:disappointed: 
Any ideas how to work around that?

Otherwise we may as well move this to 2.5

daschuer commented on src/widget/wknobcomposed.cpp:
```diff
@@ -116,6 +114,26 @@ void WKnobComposed::onConnectedControlChanged(double dParameter, double dValue)
 }
 
+void WKnobComposed::setDefaultAngleFromParameterOrReset(std::optional<double> parameter) {
+    if (!parameter.has_value() || parameter.value() < 0 || parameter.value() > 1) {
+        m_defaultAngle = std::nullopt;
+    } else {
+        m_defaultAngle = std::lerp(m_dMinAngle, m_dMaxAngle, parameter.value());
```
you may try 
```suggestion
    if (!parameter.has_value() || *parameter < 0 || *parameter > 1) {
        m_defaultAngle = std::nullopt;
    } else {
        m_defaultAngle = std::lerp(m_dMinAngle, m_dMaxAngle, *parameter);
```

ronso0 replied:
okay, great, that did the trick!
and I added a TODO as @Swiftb0y suggested.

Swiftb0y commented:
>  Any ideas how to work around that?

No idea. Just AppleClang being completely outdated as usual... 

> Otherwise we may as well move this to 2.5

Yeah. maybe leave a TODO there so the next time someone stumbles over this, they can do the trivial refactoring.

ronso0 committed:
WKnobComposed: more std::lerp

ronso0 commented:
Alright, CI all green.

Please take a look, then I'll squash the fixups (had to fixup since the commit introducing std::optional couldn't be built).

Swiftb0y commented on src/widget/wknobcomposed.cpp:
```diff
@@ -115,6 +115,8 @@ void WKnobComposed::onConnectedControlChanged(double dParameter, double dValue)
 }
 
 void WKnobComposed::setDefaultAngleFromParameterOrReset(std::optional<double> parameter) {
+    // TODO(xxx) Use m_defaultAngle->value() as soon as
+    // AppleClang 10.13+ is used.
```
ideally this just used the monadic operations from C++23 by then :wink:

ronso0 commented:
Am I done here, all reviews addressed?

Swiftb0y reviewed (commented):
from my side, yes. almost

ronso0 commented on src/effects/effectparameterslotbase.h:
```diff
@@ -50,6 +51,10 @@ class EffectParameterSlotBase : public QObject {
 
     virtual void setParameter(double value) = 0;
 
+    virtual std::optional<double> neutralPointOnScale() const {
+        return std::nullopt;
+    }
```
Then we'd have to override it for `EffectButtonParameterSlot`, just for the _possible_ case that the neutral point is requested for a button parameter?
That would otherwise return the unchanged default 0. Though, in that case (controlling a button parameter with a knob widget) a lot else wouldn't work quite as desired.
I don't mind..

Wdyt?

Swiftb0y replied:
> Then we'd have to override it for EffectButtonParameterSlot, just for the possible case that the neutral point is requested for a button parameter?

Well if that makes sense for a button parameter, we should make sure the value also makes sense. If it doesn't make sense (which is what you are suggesting, right?) then we can also let the return value be nonsensical. 

The advantage of implementing it in the base class is that we can avoid virtual dispatch. 

I think the primary issue here is that `neutralPointOnScale` only makes sense on knobs, but the existing code infrastructure requires it to be implemented on the base-class. Is that correct?

ronso0 replied:
Alright, implemented as you suggested. In the unlikely button case it'll return the default 0.

ronso0 committed:
effects: move neutralPointOnScale() to effectparameterslotbase.h

Swiftb0y reviewed (approved):
LGTM. sorry for having forgotten about this.

Swiftb0y commented:
Thank you

ronso0 commented:
> LGTM. sorry for having forgotten about this.

Never mind, there was an unaddressed review and I've pushed this back until now.
Thanks for your review!
